storage-proofs/src/layered_drgporep.rs

```diff
@@ -433,6 +451,50 @@ pub trait Layers {
                             // If we panic anywhere in this closure, thread.join() below will receive an error —
                             // so it is safe to unwrap.
                             let graph = transfer_rx.recv().unwrap();
+
+                            #[cfg(feature = "disk-trees")]
+                            let tree_d = {
+                                let tree_dir =
+                                    &settings::SETTINGS.lock().unwrap().replicated_trees_dir;
+                                // We should always be able to get this configuration
+                                // variable (at least as an empty string).
+
+                                if tree_dir.is_empty() {
+                                    // Signal `merkle_tree_path` to create a temporary file.
+                                    // FIXME: duplicating `merkle_tree_path` to avoid the
+                                    //  "temporary value dropped while borrowed" (because we
+                                    //  were creating a temporary `PathBuf` below).
+                                    graph.merkle_tree_path(&data_copy, None).unwrap()
+                                } else {
+                                    // Try to create `tree_dir`, ignore the error if `AlreadyExists`.
+                                    if let Some(create_error) = fs::create_dir(&tree_dir).err() {
+                                        if create_error.kind() != io::ErrorKind::AlreadyExists {
+                                            panic!(create_error);
+                                        }
+                                    }
+
+                                    let tree_d = graph
+                                        .merkle_tree_path(
+                                            &data_copy,
+                                            Some(&PathBuf::from(tree_dir).join(format!(
+                                                "tree-{}-{}",
+                                                layer,
+                                                rand::random::<u32>()
+                                            ))),
+                                        )
+                                        .unwrap();
```

Comment:
is it safe to `unwrap` here?

Reply:
(I see we did this before too.. )

Reply:
(to this and previous comment) the `scope` is set to return `()` (empty?) errors

https://github.com/filecoin-project/rust-fil-proofs/blob/1b3b2641e52b7742fe5ec8e404bfb326b751a2a0/storage-proofs/src/layered_drgporep.rs#L421

should I change that?

Reply:
that is not empty errors, that is just empty success value, so you can still return errors regularly

Reply:
this is a type alias defined in `errors.rs` as `type Result<T> = std::result::Result<T, Error>`

Reply:
@dignifiedquire I'm trying to return (https://github.com/filecoin-project/rust-fil-proofs/commit/616dc80619476380237074143380bf2026539708) an error but getting

```
error[E0308]: mismatched types
   --> storage-proofs/src/layered_drgporep.rs:451:59
    |
451 |                           let thread = scope.spawn(move |_| {
    |  ___________________________________________________________^
452 | |                             // If we panic anywhere in this closure, thread.join() below will receive an error —
453 | |                             // so it is safe to unwrap.
454 | |                             let graph = transfer_rx.recv().unwrap();
...   |
502 | |                             return_channel.send((layer, tree_d)).unwrap();
503 | |                         });
    | |_________________________^ expected enum `error::Error`, found ()
    |
    = note: expected type `error::Error`
               found type `()`

error: aborting due to previous error
```

I've upset Rust expectations of what is being returned but not sure how.

Reply:
Returning an `Error` at the end of the closure fixes it but I can't seem to return an `Ok` and I thought that the empty `()` signaled there's no need to return anything.

Reply:
The conditional compilation might be the problem here, I think I need to be consistent across both branches.

Reply:
Using wrong syntax, it was `Err(Error::)`.